transport/network/src/ping.rs

```diff
@@ -77,61 +78,111 @@ pub struct PingConfig {
     pub timeout: std::time::Duration, // `Duration` -> should be in millis,
 }
 
-/// Trait for the ping operation itself.
-#[async_trait]
-pub trait Pinging {
-    async fn ping(&mut self, peers: Vec<PeerId>);
+pub type PingQueryResult = std::result::Result<(std::time::Duration, String), ()>;
+
+#[derive(Debug)]
+pub struct PingQueryReplier {
+    notifier: futures::channel::oneshot::Sender<PingQueryResult>,
+    challenge: Box<(u64, ControlMessage)>,
+}
+
+impl PingQueryReplier {
+    pub fn new(notifier: futures::channel::oneshot::Sender<PingQueryResult>) -> Self {
+        Self {
+            notifier,
+            challenge: Box::new((
+                current_time().as_unix_timestamp().as_millis() as u64,
+                ControlMessage::generate_ping_request(),
+            )),
+        }
+    }
+
+    pub fn challenge(&self) -> ControlMessage {
+        self.challenge.1.clone()
+    }
+
+    pub fn notify(self, pong: ControlMessage, version: String) {
+        let timed_result = if ControlMessage::validate_pong_response(&self.challenge.1, &pong).is_ok() {
+            let unidirectional_latency = current_time()
+                .as_unix_timestamp()
+                .saturating_sub(std::time::Duration::from_millis(self.challenge.0))
+                .div(2u32);
+            Ok((unidirectional_latency, version))
+        } else {
+            Err(())
+        };
+
+        if self.notifier.send(timed_result).is_err() {
+            warn!("Failed to notify the ping query result due to timeout");
+        }
+    }
+}
+
+#[tracing::instrument(level = "trace")]
+pub fn to_active_ping(
+    peer: PeerId,
+    sender: HeartbeatSendPingTx,
+    timeout: std::time::Duration,
+) -> impl std::future::Future<Output = (PeerId, std::result::Result<std::time::Duration, ()>, String)> {
+    let (tx, rx) = futures::channel::oneshot::channel::<PingQueryResult>();
+    let replier = PingQueryReplier::new(tx);
+
+    if let Err(e) = sender.unbounded_send((peer, replier)) {
+        warn!("Failed to initiate a ping request to '{peer}': {e}");
+    }
+
+    async move {
+        match timeout_fut(timeout, rx).await {
+            Ok(Ok(Ok((latency, version)))) => {
```

Comment:
Can you document the triple wrapping?

Reply:
fallible timeout operation, fallible channel operation, fallible ping operation.